04cpp3/sec02Templates.md

```diff
@@ -0,0 +1,351 @@
+
+## Organising and Compiling Code with Templates
+
+Since templates create concrete classes and functions at compile time, the compiler needs to have access to the template definition and the argument(s) for which it needs to be instantiated together at compile time. 
+
+Well organised C++ code typically organises code into declarations and implementations:
+- Declarations are contained in header files, usually ending in `.h` or `.hpp`. (Sometimes `.h` is used to distinguish header files intended for use with C and `.hpp` for header files which are only compatible with C++, but this is just a convention and not uniformly applied.) These are typically there to declare classes, and the variables and functions that they possess, and/or the variables and functions contained in a particular namespace. Function declarations just contain the name and signature (input types and return type) of the function, but not the actual code that it executes. When writing code which uses a class we usually just include the header file in our code to include the declaration: as long as we know the interface for the class we can compile code down to an object file which interfaces with that class. The final code can be created by _linking_ with the class object file which defines the actual implementation that needs to be executed. 
+- Implementations are contained in source files, usually ending in `.cpp` for C++ files. These contain the actual code which is executed by the functions declared in the header. (There may also be functions declared in the source file which aren't in the header if they're only locally needed and therefore don't need to be included elsewhere.) This can then be compiled down to an object file for linking with other object files which interface with each other. 
+
+We have to be a little careful when working with this model in the case of templates. We'll explore this using a function template, although the considerations are the same for a class template as well. Let's declare a function in a header file, and write two source files: one which implements the function in the header, and one which makes use of this function. 
+
+First the declaration in `declaration.hpp`:
+```cpp
+#ifndef DECLARATION_HPP 
+#define DECLARATION_HPP
+
+namespace utilFunctions
+{
+    int add(int, int);
+}
+
+#endif
+```
+- It's usually a good idea to put functions and variables which are defined in a header but not part of a class inside a namespace, to avoid potential name clashes. 
+- The `#ifndef`, `#define` and `#endif` are pre-processor directives. This pattern is called an "include guard": it means that the file's contents will be ignored if it has already been included somewhere else in the same compilation unit, so that the contents are not declared twice (which would cause an error). 
```

Comment:
Could mention that `#pragma once` is also widely used but technically not part of ISO C++.